compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -409,6 +409,78 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             err.downgrade_to_delayed_bug();
         }
 
+        if let (ty::Adt(adt_def, _), SelfSource::QPath(_)) = (rcvr_ty.kind(), source) {
+            let mut items = self
+                .tcx
+                .inherent_impls(adt_def.did())
+                .iter()
+                .flat_map(|i| self.tcx.associated_items(i).in_definition_order())
+                .filter(|item| {
+                    matches!(item.kind, ty::AssocKind::Fn) && !item.fn_has_self_parameter
+                })
+                .filter_map(|item| {
+                    let ret_ty = self.tcx.fn_sig(item.def_id).skip_binder().output().skip_binder();
+                    let ty::Adt(def, args) = ret_ty.kind() else {
+                        return None;
+                    };
+                    if ![
```

Comment:
Would appreciate if you left comments explaining some of this logic, so that later users fixing/extending it will understand why it does what it does.

e.g. this avoids `Result` and `Option` why exactly? I can guess, but useful for posterity.

Reply:
Done. Pointing to this made me realize that there was a logic bug: it was *only* looking at returns of Option and Result, but not returns of Self. Fixed.